presto-hive/src/main/java/com/facebook/presto/hive/SyncPartitionMetadataProcedure.java

```diff
@@ -0,0 +1,258 @@
+    }
+
+    private static void syncPartitions(
+            Set<String> partitionsToAdd,
+            Set<String> partitionsToDrop,
+            SyncMode syncMode,
+            SemiTransactionalHiveMetastore metastore,
+            ConnectorSession session,
+            Table table)
+    {
+        if (syncMode == SyncMode.ADD || syncMode == SyncMode.FULL) {
+            addPartitions(metastore, session, table, partitionsToAdd);
+        }
+        if (syncMode == SyncMode.DROP || syncMode == SyncMode.FULL) {
+            dropPartitions(metastore, session, table, partitionsToDrop);
+        }
+        metastore.commit();
+    }
+
+    private static void addPartitions(
+            SemiTransactionalHiveMetastore metastore,
+            ConnectorSession session,
+            Table table,
+            Set<String> partitions)
+    {
+        for (String name : partitions) {
+            metastore.addPartition(
+                    session,
+                    table.getDatabaseName(),
+                    table.getTableName(),
+                    table.getStorage().getLocation(),
+                    false,
+                    buildPartitionObject(session, table, name),
+                    new Path(table.getStorage().getLocation(), name),
+                    PartitionStatistics.empty());
+        }
+    }
+
+    private static void dropPartitions(
+            SemiTransactionalHiveMetastore metastore,
+            ConnectorSession session,
+            Table table,
+            Set<String> partitions)
+    {
+        for (String name : partitions) {
+            metastore.dropPartition(
+                    session,
+                    table.getDatabaseName(),
+                    table.getTableName(),
+                    table.getStorage().getLocation(),
+                    extractPartitionValues(name));
+        }
+    }
+
+    private static Partition buildPartitionObject(ConnectorSession session, Table table, String partitionName)
+    {
+        return Partition.builder()
+                .setDatabaseName(table.getDatabaseName())
+                .setTableName(table.getTableName())
+                .setColumns(table.getDataColumns())
+                .setValues(extractPartitionValues(partitionName))
+                .setParameters(ImmutableMap.of(PRESTO_QUERY_ID_NAME, session.getQueryId()))
+                .withStorage(storage -> storage
+                        .setStorageFormat(table.getStorage().getStorageFormat())
+                        .setLocation(new Path(table.getStorage().getLocation(), partitionName).toString())
+                        .setBucketProperty(table.getStorage().getBucketProperty())
+                        .setSerdeParameters(table.getStorage().getSerdeParameters()))
+                .build();
+    }
+
+    private static SyncMode toSyncMode(String mode)
+    {
+        try {
+            return SyncMode.valueOf(mode.toUpperCase(ENGLISH));
+        }
+        catch (IllegalArgumentException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)